otlp_json2/__init__.py

```diff
@@ -0,0 +1,173 @@
+from __future__ import annotations
+
+import json
+from collections.abc import Mapping, Sequence
+from typing import Any, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing_extensions import TypeAlias
+
+    from opentelemetry.trace import Link
+    from opentelemetry._logs import LogRecord
+    from opentelemetry.sdk.trace import ReadableSpan, Event
+    from opentelemetry.sdk.resources import Resource
+    from opentelemetry.sdk.util.instrumentation import InstrumentationScope
+    from opentelemetry.trace.status import Status
+
+    _LEAF_VALUE: TypeAlias = "str | int | float | bool"  # TODO: confirm
+    _VALUE: TypeAlias = "_LEAF_VALUE | Sequence[_LEAF_VALUE]"
+
+
+CONTENT_TYPE = "application/json"
+
+
+_VALUE_TYPES = {
+    # NOTE: order matters, for isinstance(True, int).
+    bool: ("boolValue", bool),
+    int: ("intValue", str),
+    float: ("doubleValue", float),
+    bytes: ("bytesValue", bytes),
+    str: ("stringValue", str),
+    Sequence: (
+        "arrayValue",
+        lambda value: {"values": [_value(e) for e in _homogeneous_array(value)]},
+    ),
+    Mapping: (
+        "kvlistValue",
+        lambda value: {"values": [{k: _value(v) for k, v in value.items()}]},
+    ),
+}
+
+
+def encode_spans(spans: Sequence[ReadableSpan]) -> bytes:
+    spans = sorted(spans, key=lambda s: (id(s.resource), id(s.instrumentation_scope)))
+    rv = {"resourceSpans": []}
+    last_rs = last_is = None
```

Comment:
Some of these names are a bit confusing. What about rv -> spans, last_rs -> last_resource, last_is -> last_scope?

Also, to simplify, why not make `spans` just the list, and then only do `{"resourceSpans": spans}` at the end?